File: C++/3011. Find if Array Can Be Sorted.cpp
class Solution {
public:
    int countSetBits(int n) {
        int count = 0;
        while(n) {
            count += (n & 1);  
            n >>= 1;          
        }
        return count;
    }
    
    bool canSortArray(vector<int>& nums) {
        int n = nums.size();
        vector<int> setBits(n);
        for(int i = 0; i < n; i++) {
            setBits[i] = countSetBits(nums[i]);
        }
        bool swapped;
        do {
            swapped = false;
            for(int i = 0; i < n-1; i++) {
                if(setBits[i] == setBits[i+1] && nums[i] > nums[i+1]) {
                    swap(nums[i], nums[i+1]);
                    swapped = true;
                }
            }
        } while(swapped);
        for(int i = 0; i < n-1; i++) {
            if(nums[i] > nums[i+1]) return false;
        }
        
        return true;
    }
};

File: JAVA/3011. Find if Array Can Be Sorted.java

class Solution {
    public boolean canSortArray(int[] nums) {
        int n=nums.length;
        boolean flag=true;
        
        for(int i=0;i<n-1;i++){
            if(nums[i]>nums[i+1]){
                flag=false;
            }
        }
        if(flag){
            return true;
        }
        for(int i=0;i<n-1;i++){
            for(int j=0;j<n-i-1;j++){
                if((nums[j]>nums[j+1]) && (getBits(nums[j])==getBits(nums[j+1]))){
                    int temp=nums[j];
                    nums[j]=nums[j+1];
                    nums[j+1]=temp;
                }
            }
        }
        for(int i=0;i<n-1;i++){
            if(nums[i]>nums[i+1]) return false;
        }
        return true;
        
    }
    
    public int getBits(int n){
        int bitCount=0;
        
        while(n!=0){
            bitCount+=(n&1);
            n=n>>1;
        }
        return bitCount;
    }
}
